nbs/20_comparison_others/py/05_02-gtex_whole_blood-sample_all_gene_pairs-mic.py

```diff
@@ -0,0 +1,271 @@
+display(OUTPUT_DIR)
+
+# %% tags=[]
+OUTPUT_FILE_TEMPLATE = str(
+    OUTPUT_DIR / (INPUT_GENE_PAIRS_FILE.name[:-4] + f"-{METHOD_NAME}.pkl")
+)
+
+display(OUTPUT_FILE_TEMPLATE)
+
+# %% [markdown] tags=[]
+# # Data
+
+# %% [markdown] tags=[]
+# ## Gene expression
+
+# %% tags=[]
+gene_expr_dict = pd.read_pickle(INPUT_GENE_EXPR_FILE).T.to_dict(orient="series")
+
+# %% tags=[]
+len(gene_expr_dict)
+
+# %% tags=[]
+gene_expr_dict[list(gene_expr_dict.keys())[0]]
+
+# %% [markdown] tags=[]
+# ## Get all sample files
+
+# %% tags=[]
+all_sample_files = []
+
+sample_id = 0
+sample_file = Path(INPUT_GENE_PAIRS_FILE_TEMPLATE.format(sample_id=sample_id))
+display(sample_file)
+
+while sample_file.exists():
+    all_sample_files.append((sample_id, sample_file))
+
+    sample_id += 1
+    sample_file = Path(INPUT_GENE_PAIRS_FILE_TEMPLATE.format(sample_id=sample_id))
+
+# %% tags=[]
+display(len(all_sample_files))
+assert len(all_sample_files) > 0
+
+# %% tags=[]
+all_sample_files[:3]
+
+
+# %% [markdown] tags=[]
+# # Compute Maximal Information Coefficient (MIC)
+
+# %% [markdown] tags=[]
+# ## Functions
+
+# %% tags=[]
+def _mic(x, y):
+    """
+    Given two arrays (x and y), it computes MIC with the default parameters.
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=DeprecationWarning)
+
+        mine = MINE(alpha=0.6, c=15, est="mic_approx")
+        mine.compute_score(x, y)
+        return mine.mic()
+
+
+# %% tags=[]
+# testing
+_mic_val = _mic(np.random.rand(10), np.random.rand(10))
+display(_mic_val)
+assert _mic_val > 0.0
+
+
+# %% tags=[]
+def _compute_mic(gene_sets: list):
+    """
+    It takes a list of gene pairs and computes MIC on all.
+    It returns a series with gene pairs as index and MIC values.
+    This function is used in concurrent.futures for parallel execution.
+    """
+    res = [
+        _mic(gene_expr_dict[gs[0]].to_numpy(), gene_expr_dict[gs[1]].to_numpy())
+        for gs in gene_sets
+    ]
+
+    return pd.Series(res, index=pd.MultiIndex.from_tuples(gene_sets))
+
+
+# %% tags=[]
+# testing
+# gene_set_key = "Clustermatch (high), Pearson (high), Spearman (high)"
+gene_set = pd.read_pickle(all_sample_files[0][1]).sample(n=10)
+display(gene_set)
+
+_res = _compute_mic(list(gene_set.itertuples(index=False)))
+display(_res.shape)
+display(_res.head())
+
+# make sure order is preserved
+assert _res.index.to_list() == list(gene_set.itertuples(index=False, name=None))
+
+# %% [markdown] tags=[]
+# ## Run
+
+# %% tags=[]
+# create chunks of gene pairs to compute in parallel
+all_chunks = []
+
+for (sample_id, sample_file) in all_sample_files:
+    gene_pairs_df = pd.read_pickle(sample_file)
+    gene_pairs_subset = list(gene_pairs_df.itertuples(index=False, name=None))
+
+    for chunk in list(chunker(gene_pairs_subset, CHUNK_SIZE)):
+        all_chunks.append((sample_id, chunk))
+
+# %% tags=[]
+len(all_chunks)
+
+# %% tags=[]
+all_chunks[:2]
```

Comment:
So an element of all_chunks is (0, [...]). Is the 0 the sample id?

Reply:
Yes, correct. I needed a way to get the sample id out of the file name in a safe manner without using regex, etc.